util/src/sync/init_once.rs

```diff
@@ -0,0 +1,296 @@
+            ",
+            any::type_name::<T>(),
+        );
+        unsafe {
+            // Safety: hahaha wheeee no rules! You can't stop meeeeee!
+            &*((*self.value.get()).as_ptr())
+        }
+    }
+}
+
+impl<T> core::ops::Deref for InitOnce<T> {
+    type Target = T;
+
+    #[track_caller]
+    #[inline]
+    fn deref(&self) -> &Self::Target {
+        match self.get() {
+            Some(t) => t,
+            None => panic!("InitOnce<{}> not yet initialized!", any::type_name::<T>(),),
+        }
+    }
+}
+
+impl<T: fmt::Debug> fmt::Debug for InitOnce<T> {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        let mut d = f.debug_struct("InitOnce");
+        d.field("type", &any::type_name::<T>());
+        match self.state.load(Ordering::Acquire) {
+            INITIALIZED => d.field("value", Deref::deref(self)).finish(),
+            INITIALIZING => d.field("value", &format_args!("<initializing>")).finish(),
+            UNINITIALIZED => d.field("value", &format_args!("<uninitialized>")).finish(),
+            state => unreachable!("unexpected state value {}, this is a bug!", state),
+        }
+    }
+}
+
+// === impl Lazy ===
+
+impl<T> Lazy<T> {
+    pub const fn new(initializer: fn() -> T) -> Self {
+        Self {
+            value: UnsafeCell::new(MaybeUninit::uninit()),
+            state: AtomicU8::new(UNINITIALIZED),
+            initializer,
+        }
+    }
+
+    /// Initialize the cell to `value`, returning an error if it has already
+    /// been initialized.
+    ///
+    /// If the cell has already been initialized, the returned error contains
+    /// the value.
+    pub fn get(&self) -> &T {
+        let state = self.state.compare_exchange(
+            UNINITIALIZED,
+            INITIALIZING,
+            Ordering::AcqRel,
+            Ordering::Acquire,
+        );
+
+        match state {
+            Err(INITIALIZED) => {
+                // Already initialized! Just return the value.
+            }
+            Err(INITIALIZING) => {
+                // Wait for the cell to be initialized
+                let mut backoff = super::Backoff::new();
+                while self.state.load(Ordering::Acquire) != INITIALIZED {
+                    backoff.spin();
+                }
+            }
+            Ok(_) => {
+                // Now we have to actually initialize the cell.
+                unsafe {
+                    *(self.value.get()) = MaybeUninit::new((self.initializer)());
+                }
+                if let Err(actual) = self.state.compare_exchange(
+                    INITIALIZING,
+                    INITIALIZED,
+                    Ordering::AcqRel,
+                    Ordering::Acquire,
+                ) {
+                    unreachable!(
+                        "Lazy<{}>: state changed while locked. This is a bug! (state={})",
+                        any::type_name::<T>(),
+                        actual
+                    );
+                }
```

Comment:
Oh, why use `compare_exchange` here? Seems like because it's a bug for it to have changed, we should just `.store(INITIALIZED, Ordering::AcqRel)`, or `.swap` if we want to inspect the old value for assertions.

Reply:
yeah this could be swap, good point!